bigchaindb/fastquery.py

```diff
@@ -0,0 +1,65 @@
+from bigchaindb.utils import output_has_owner
+from bigchaindb.backend import query
+from bigchaindb.common.transaction import TransactionLink
+
+
+class FastQuery:
+    """
+    Database queries that join on block results from a single node.
+
+    * Votes are not validated for security (security is a replication concern)
+    * Votes come from only one node, and as such, non-byzantine fault tolerance
+      is reduced.
+
+    Previously, to consider the status of a block, all votes for that block
+    were retrieved and the election results were counted. This meant that a
+    faulty node may still have been able to obtain a correct election result.
+    However, from the point of view of a client, it is still neccesary to
+    query multiple nodes to insure against getting an incorrect response from
+    a byzantine node.
+    """
+    def __init__(self, connection, me):
+        self.connection = connection
+        self.me = me
+
+    def filter_block_ids(self, block_ids, include_undecided=True):
+        """
+        Given block ids, filter the invalid blocks.
+        """
+        block_ids = list(set(block_ids))
+        votes = query.get_votes_for_blocks_by_voter(
+                    self.connection, block_ids, self.me)
+        votes = {v['vote']['voting_for_block']: v['vote']['is_block_valid'] for v in votes}
+        return [b for b in block_ids if votes.get(b, include_undecided)]
+
+    def filter_valid_blocks(self, blocks, key=lambda b: b[0]):
+        """
+        Given things with block ids, remove the invalid ones.
+        """
+        blocks = list(blocks)
+        valid_block_ids = set(self.filter_block_ids(key(b) for b in blocks))
```

Comment:
I think calling `filter_block_ids` with it's default parameter `include_undecided=True` (which should not even be the option IMO anyways) is potentially dangerous.

**Requirement:**
https://docs.bigchaindb.com/projects/server/en/latest/http-client-server-api.html#transaction-outputs doesn't state whether or not it returns transactions from invalid blocks. We should however have the endpoint return only transactions from valid blocks applying a best-effort & non-embarrassing performance strategy upon retrieval.


I'd either change `self.filter_block_ids(key(b)` it to include `include_undecided=False`, or remove the parameter from `filter_block_ids`.